torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp

```diff
@@ -285,6 +285,154 @@ void ThreadPredicateMap::updateBitSet(const Expr* expr) {
   }
 }
 
+namespace {
+
+//! A simple backward data flow pass:
+//!  This pass propagates information backward to annotate "redundant use
+//!  chain"'s.
+//! The reason this is needed is that if we have a chain of register-to-register
+//! ops
+//!  that begins with a redundant shared/global mem write and ends with an op,
+//!  typically
+//! a broadcast, that non-redundantly uses the result, we'd need to insert a
+//! sync at the begining of the register-to-register chain.
+//!
+//! The propagation rule is summarized as follows:
+//!
+//!   Shared TV val:
+//!      Reset all block redundant info to its own redundant write info
+//!      Backpropagate grid redundant info
+//!   Global TV val:
+//!      Reset all redundant info to its own redundant write info
+//!   Local Tv val:
+//!      Backpropagate all redundant info
+//!   Exprs:
+//!      Propagate redundant info backwards from outputs to inputs:
+//!        For each parallel type,
+//!          The parallel type is redundantly used in the expr input
+//!          only if all of the outputs redundantly use the same type.
+class RedundantUseAnalysis : BackwardVisitor {
+ public:
+  RedundantUseAnalysis(Fusion* fusion, const ThreadPredicateMap& pred_map)
+      : fusion_(fusion), pred_map_(pred_map) {
+    traverseFrom(fusion, fusion->terminatingMathVals());
+  }
+
+  c10::optional<ParallelTypeBitmap> getRedundantUseBitMap(
+      const TensorView* tv) {
+    auto redundant_map_it = redundant_tv_use_map_.find(tv);
+    if (redundant_map_it == redundant_tv_use_map_.end()) {
+      return c10::nullopt;
+    } else {
+      return redundant_map_it->second;
+    }
+  }
+
+ private:
+  using BackwardVisitor::handle;
+
+  void handle(TensorView* tv) final {
```

Comment:
If I read this function correctly, it seems that the `redundant_types` of `tv`, when it's shared or local, doesn't matter and `redundant_tv_use_map_[tv]` is solely determined by the consumers of `tv`. Is it correct?

Reply:
I think that's correct. 

`redundant_types` is the same definition as before.

`redundant_tv_use_map_` is the new information that is derived from the `redundant_types` of the consumer chains.

Reply:
So, a TV can have `redundant_tv_use` parallel types even if it doesn't have anything in `redundant_types`?

Reply:
Yes. The `redundant write` and `redundant read` are two pieces of info that complement each other and each could be useful in different scenarios.

In the context of this PR, we can remove the sync when the `redundant read` and `redundant write` are fully overlapped.

Reply:
Hmm, I'm still confused. In the header file, it says:

```
Tracking use chain of redundant writes:
```

If `a TV can have redundant_tv_use parallel types even if it doesn't have anything in redundant_types` is true, it seems it tracks use chains of non-redundant writes as well. What am I missing?

Reply:
Oh that looks like confusing wording, would change it to `use chains that end with redundant writes`. 

The information it conveys is different from `redundant writes` and it is totally valid for a tv that isn't redundantly written but say only threads with `tidx==0` actually use it. Scenario like this is modeled in the analysis though it's not really used in this PR.

Reply:
OK, I guess I'm slowly getting this, but calling it "redundant reads" still sounds confusing. For example, if a tensor is never used but is just output, and it is redundantly written, the algorithm would make it as also redundantly read, right?

Could you add some unit tests just for this pass? Some small tests demonstrating its intentions as well as exercising corner cases would be really helpful to get the gist of this analysis.

Reply:
Was reviewing the code again, and now I think the major confusion comes from the way the traversal is viewed. The info in the predicateInfo is conceptually one iteration earlier than `redundant type`. So `redundant consumer` actually meant `redundant as a consumer`. So in your example is an output is redundantly written, then it's producer can skip syncs since the output is `redundant as a consumer`. 

I guess the reason it was done this way originally was to use it as a drop in fix in sync info pass, which actually does look into immediate consumers, where this info was applied.

Now I agree it's more confusing this way and I will just add a final iteration to aggregate for each tv so that the `redundant use` and `redundant write` look conceptually more symmetric. 

Also definitely I could add more unit tests on the info pass.